pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -90,117 +100,255 @@ func (c *ClusterMemberController) reconcileMembers(ctx context.Context, recorder
+		if isURLMappedToMember(peerURL, members) {
+			continue
+		}
+
+		// If the Machine API is unavailable then we can't evaluate the machine's
+		// deletion timestamp or deletion hooks so we scale it up without any considerations
+		// for vertical scaling
+		// This allows UPI clusters without a machine API to scale up
+		// Check machine API availability
+		isMachineAPIFunctional, err := c.machineAPIChecker.IsFunctional()
+		if err != nil {
+			return "", fmt.Errorf("failed to determine Machine API availability: %v", err)
+		}
+		if !isMachineAPIFunctional {
+			return peerURL, nil
+		}
+
+		// If the Machine API is functional we should ignore the machine
+		// if it is pending deletion to avoid re-adding previously removed members
+		internalIP, err := c.getInternalIPForNode(node)
+		if err != nil {
+			return "", fmt.Errorf("failed to get internal IP for node: %v", err)
+		}
+		machine, err := ceohelpers.FindMachineByNodeInternalIP(internalIP, c.masterMachineSelector, c.masterMachineLister)
+		if err != nil {
+			return "", fmt.Errorf("failed to get machine for node (%s): %v", node.Name, err)
+		}
+		if machine == nil {
+			klog.V(2).Infof("Ignoring node (%s) for scale-up: no Machine found referencing this node's internal IP (%v)", node.Name, internalIP)
+			continue
+		}
+		if machine.DeletionTimestamp != nil {
+			// Ignore this node since its member previously was, or is in the process of being removed
+			klog.V(4).Infof("Ignoring node (%s) for scale-up since its machine (%s) is pending deletion", node.Name, machine.Name)
+			continue
+		}
+
+		// Wait until the machine has a deletion hook present before adding it as a member
+		if !ceohelpers.HasMachineDeletionHook(machine) {
+			continue
+		}
+
+		// Wait until the etcd pod is running but not ready on the node which indicates
+		// it is waiting to be added to the cluster membership
+		// Note: Checking the pod before adding the member ensures no two unstarted members are added in parallel.
+		// If multiple new nodes are added as members in parallel they will fail with
+		// a discovery error ("error validating peerURLs: ... member count is unequal") because
+		// the unstarted members won't include each other in their ETCD_INITIAL_CLUSTER env.
+		podName := fmt.Sprintf("etcd-%v", node.Name)
+		runningNotReady, err := c.isEtcdContainerRunningNotReady(podName)
+		if err != nil {
+			return "", fmt.Errorf("failed to check if pod is running and not ready: %v", err)
+		}
+		if !runningNotReady {
+			continue
+		}
+
+		// We have a valid new node with a running etcd pod to add as a new member
+		return peerURL, nil
+	}
+	return "", nil
+}
+
+// ensureEtcdLearnerPromotion checks for all learner members in the cluster membership and tries
+// to promote them if they have the required machine deletion hooks present
+func (c *ClusterMemberController) ensureEtcdLearnerPromotion(ctx context.Context, recorder events.Recorder) error {
+	members, err := c.etcdClient.MemberList(ctx)
```

Comment:
We could rely on the information stored in the cache to indicate work. That would make this controller cheap in the steady-state.

Reply:
similar to https://github.com/openshift/cluster-etcd-operator/pull/758#discussion_r855056766, wdyt?